File: WdtcCore/src/main/java/org/wdt/wdtc/core/manger/TaskManger.java
package org.wdt.wdtc.core.manger;

import com.google.gson.JsonObject;
import org.apache.log4j.Logger;
import org.wdt.utils.io.FileUtils;
import org.wdt.utils.io.IOUtils;
import org.wdt.wdtc.core.download.game.DownloadVersionGameFile;
import org.wdt.wdtc.core.utils.DownloadUtils;
import org.wdt.wdtc.core.utils.WdtcLogger;
import org.wdt.wdtc.core.utils.gson.JSONUtils;

import java.io.File;
import java.io.IOException;

import static java.util.Objects.requireNonNull;

public class TaskManger {
  private static final Logger logmaker = WdtcLogger.getLogger(TaskManger.class);

  public static void ckeckVMConfig() {
    if (VMManger.isDebug()) {
      System.setProperty(VMManger.CONFIG_PATH, "./");
    }
    if (System.getProperty(VMManger.CONFIG_PATH) == null) {
      System.setProperty(VMManger.CONFIG_PATH, System.getProperty("user.home"));
    }
  }

  public static void runStartUpTask() throws IOException {
    FileUtils.delete(DownloadUtils.StopProcess);
    FileUtils.writeStringToFile(new File(FileManger.getWdtcConfig(), "assets/readme.txt"),
        IOUtils.toString(requireNonNull(SettingManger.class.getResourceAsStream("/assets/readme.txt"))));
    FileUtils.createDirectories(FileManger.getWdtcCache());
    if (FileUtils.isFileNotExists(FileManger.getUserListFile())) {
      JSONUtils.writeObjectToJsonFile(FileManger.getUserListFile(), new JsonObject());
    }
    if (FileUtils.isFileNotExists(FileManger.getSettingFile())) {
      JSONUtils.writeObjectToJsonFile(FileManger.getSettingFile(), new SettingManger.Setting());
    }
    String LlbmpipeLoader = "https://maven.aliyun.com/repository/public/org/glavo/llvmpipe-loader/1.0/llvmpipe-loader-1.0.jar";
    if (FileUtils.isFileNotExists(FileManger.getLlbmpipeLoader())) {
      DownloadUtils.StartDownloadTask(LlbmpipeLoader, FileManger.getLlbmpipeLoader());
    }
    if (FileUtils.isFileNotExists(FileManger.getVersionManifestFile())) {
      DownloadVersionGameFile.DownloadVersionManifestJsonFile();
    }
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/manger/TaskManger.kt
package org.wdt.wdtc.core.manger

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import org.wdt.utils.gson.writeObjectToFile
import org.wdt.utils.io.IOUtils
import org.wdt.utils.io.createDirectories
import org.wdt.utils.io.deleteDirectory
import org.wdt.utils.io.isFileNotExists
import org.wdt.wdtc.core.auth.UsersList
import org.wdt.wdtc.core.auth.serializeUsersListGsonBuilder
import org.wdt.wdtc.core.download.game.DownloadVersionGameFile
import org.wdt.wdtc.core.game.VersionsList
import org.wdt.wdtc.core.game.ckeckVersionsList
import org.wdt.wdtc.core.utils.*
import org.wdt.wdtc.core.utils.gson.serializeVersionGson
import org.wdt.wdtc.core.utils.gson.serializeVersionsListGson
import java.io.File

fun ckeckRunEnvironment() {
	if (isMacos) {
		error("Wdtc cannot run on macos!")
	}
	if (isLinux) {
		logmaker.warning("Wdtc not recommended to run on Linux")
	}
	if (wdtcConfig.isFileNotExists()) {
		logmaker.info("First run, init config")
	}
}

fun createNeedDirectories() {
	wdtcConfig.createDirectories()
	wdtcCache.createDirectories()
	wdtcUser.createDirectories()
	wdtcDependenciesDirectory.createDirectories()
	userAsste.createDirectories()
}

fun ckeckVMConfig() {
	if (isDebug) System.setProperty(CONFIG_PATH, "./")
	if (System.getProperty(CONFIG_PATH) == null) {
		System.setProperty(CONFIG_PATH, System.getProperty("user.home"))
	}
}


fun runStartUpTask() = runBlocking(Dispatchers.IO) {
	IOUtils.copy(
		getResourceAsStream("/assets/readme.txt"), File(wdtcConfig, "readme.txt").outputStream()
	)
	userListFile.run {
		if (isFileNotExists()) {
			writeObjectToFile(UsersList(), serializeUsersListGsonBuilder)
		}
	}
	settingFile.run {
		if (isFileNotExists()) {
			writeObjectToFile(Setting(), serializeVersionGson)
		}
	}
	versionsJson.run {
		if (isFileNotExists()) {
			writeObjectToFile(VersionsList(), serializeVersionsListGson)
		}
	}
	launch {
		writeConfigJsonToAllVersion()
		ckeckVersionsList()
	}
	
	launch {
		llbmpipeLoader.let { file ->
			"https://maven.aliyun.com/repository/public/org/glavo/llvmpipe-loader/1.0/llvmpipe-loader-1.0.jar".toURL().let {
				if (file.compareFile(it.toFileData())) {
					startDownloadTask(it, file)
				}
			}
		}
	}
	if (versionManifestFile.isFileNotExists()) {
		launch { DownloadVersionGameFile.startDownloadVersionManifestJsonFile() }
	}
	
}

fun removeConfigDirectory(boolean: Boolean) {
	if (boolean) {
		wdtcConfig.deleteDirectory()
		GameDirectoryManger(currentSetting.defaultGamePath).gameVersionList.run {
			forEach {
				it.versionConfigFile.delete()
			}
		}
	}
}

open class TaskManger(
	val actionName: String,
	val actionKind: TaskKind = TaskKind.FUNCTION,
	var coroutinesAction: Job? = null,
	var action: (() -> Unit)? = null
) {
	open fun start() {
		if (actionKind == TaskKind.FUNCTION) {
			action.noNull().invoke()
		} else {
			coroutinesAction.noNull().start()
		}
	}
	
	override fun toString(): String {
		return "TaskManger(actionName='$actionName', actionKind=$actionKind, coroutinesAction=$coroutinesAction, action=$action)"
	}
	
	
}

enum class TaskKind {
	FUNCTION, COROUTINES
}